## Safety and Privacy with LangChain

In this notebook we will use our privacy (PII entity recognizer and anonymizer) and toxic text classifier `tensor-trek/distilbert-toxicity-classifier` along with LangChain to implement checks of the text going into our LLM and text generated by the LLM. For this example we will use HuggingFace Hub LLM with LangChain and a custom `PrivacyAndSafetyChain` chain that implements the two checks.


Let's install some dependencies first
- You will need `transformers`, PyTorch, `langchain`, `presidio-analyzer`, `presidio-anonymizer`, and `spacy` libraries
- You will also need the spacy `en_core_web_lg` model. You can also work with `en_core_web_md` model here.

In [ ]:
!pip install pip-system-certs -q
!pip install openai langchain==0.1 transformers torch torchvision -q
!pip install presidio-analyzer presidio-anonymizer spacy huggingface-hub -q
!python -m spacy download en_core_web_lg

/Users/idriouich/anaconda3/lib/python3.10/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'raw.githubusercontent.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━╸━━━━━━━━━━━ 415.0/587.7 MB 7.7 MB/s eta 0:00:23

In [40]:
import os
os.environ['CURL_CA_BUNDLE'] = ''
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
pd.set_option('display.max_colwidth', None)
import openai
from dotenv import load_dotenv, find_dotenv

In [41]:
_ = load_dotenv(find_dotenv())  # read local .env file

openai_api_key = os.getenv("AZURE_OPENAI_KEY")
openai_api_base = os.getenv("AZURE_OPENAI_ENDPOINT") 
openai_api_type = 'azure'
openai_api_version = '2023-08-01-preview' 
model_deployment_name = "gpt-35-turbo-0613"
embedding_deployment_name = "gpt-35-turbo-0613"

## Import the `PrivacyAndSafetyChain` custom chain

The directory `PrivacyAndSafety` contains files that implements the custom Chain.
- The `privacy_and_safety.py` file contains a Subclass of the base LangChain `Chain` class
- The `check.py` file contains the actual toxic text classification and PII entity detection and anonymization

Let's import and initialize `PrivacyAndSafetyChain` first.

##  Configuration for `PrivacyAndSafetyChain`

We can customize the behavior of `PrivacyAndSafetyChain` via the following parameters

- `pii_mask_character`, the character used to perform anonymization of PII entities. Default is `*`
- `pii_labels` if you wish to specify a specific list of PII entity types, then a list of entity types. For a full list of PII entity labels refer [Presidio supported entities](https://microsoft.github.io/presidio/supported_entities/). Defaults to ALL entities.
- `fail_on_pii` a boolean flag which will make the chain fail if PII is detected. Defaults to `False`.
- `pii_threshold` the confidence score threshold for PII entity recognition. Defaults to 50%
- `toxicity_threshold` the confidence score threshold for toxicity classification. Defaults to 80%

In [42]:
from chains import PrivacyAndSafetyChain

safety_privacy = PrivacyAndSafetyChain(verbose=True, 
                                    pii_mask_character="#",
                                    pii_labels = ["PHONE_NUMBER", "EMAIL_ADDRESS", "PERSON", "US_SSN"],
                                    fail_on_pii = True,
                                    pii_threshold = 0.5,
                                    toxicity_threshold = 0.6)

In [43]:
from openai import AzureOpenAI

# gets the API Key from environment variable AZURE_OPENAI_API_KEY
client = AzureOpenAI(
    # https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#rest-api-versioning
    api_version=openai_api_version,
    api_key = openai_api_key,
    # https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource?pivots=web-portal#create-a-resource
    azure_endpoint=openai_api_base,
)


# this is the name of the deployments you created in the Azure portal within the above resource
from typing import List, Dict

def get_chat_with_conversation(
        text,
        temperature: float = 0.2,
        **model_kwargs
) -> str:
    try:
        
        messages = [
            {"role": "system", "content": '"""'+ str(text) + '"""'}
        ]
        response = client.chat.completions.create(model=model_deployment_name,
                                                  messages=messages)
 
        return response.choices[0].message.content
    except openai.OpenAIError as e: # this is the base class of any openai exception
        print(f"The call to the Chat Completion API failed as a consequence "
              f"of the following exception: {e}")

        
def user_request():
    # Take request
    request = input("\nEnter an instruction"
                    "(or 'quit'):")
    if request.lower() == "quit":
        raise KeyboardInterrupt()
    return request

def user_reply_success(request,response):
    # Create and print user reply
    reply = f"{request}:\n{response}"
    print(reply)

# LangCahin Expression Language (LCEL)

In [44]:
import ipywidgets as widgets
display(widgets.HTML(f'<img src="{"../assets/LCEL1.png"}" width="950" align="center">'))
display(widgets.HTML(f'<img src="{"../assets/LCEL2.png"}" width="950" align="center">'))

HTML(value='<img src="../assets/LCEL1.png" width="950" align="center">')

HTML(value='<img src="../assets/LCEL2.png" width="950" align="center">')

## Now, let's wrap it up together and create our app using LangChain Expression Language (LCEL).

## The chain is now containing :



In [45]:
import ipywidgets as widgets
display(widgets.HTML(f'<img src="{"../assets/PrivacySafetyChain.png"}" width="950" align="center">'))


HTML(value='<img src="../assets/PrivacySafetyChain.png" width="950" align="center">')

In [46]:
from langchain import HuggingFaceHub
from langchain import PromptTemplate, LLMChain

template = """{question}"""

prompt = PromptTemplate(template=template, input_variables=["question"])

chain = (
    prompt 
    | safety_privacy 
    | {"input": (lambda x: x['output'] )}
    | get_chat_with_conversation
    | safety_privacy 
)


while True : 
    try:
        input_user = user_request()
        response = chain.invoke({"question": input_user})
    except Exception as e:
        print(str(e))
    else:

        print("\n The answer is: \n", response['output'])


Enter an instruction(or 'quit'):I hate my coworkers, can you help me get them fired?


> Entering new PrivacyAndSafetyChain chain...
Running PrivacyAndSafetyChain...
Checking for Toxic content...
Toxic content found in text. Stopping...
Toxic content found in text. Stopping...


Enter an instruction(or 'quit'):What is John Doe's email address, phone number and SSN from the following text? John Doe, a resident of 1234 Elm Street in Springfield, recently celebrated his birthday on January 1st. Turning 43 this year, John reflected on the years gone by. He often shares memories of his younger days with his close friends through calls on his phone, (555) 123-4567. Meanwhile, during a casual evening, he received an email at johndoe@example.com reminding him of an old acquaintance's reunion. As he navigated through some old documents, he stumbled upon a paper that listed his SSN as 338-12-6789, reminding him to store it in a safer place."


> Entering new PrivacyAndSafetyChain chain...
Runni

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:19                                                                                   │
│                                                                                                  │
│   16                                                                                             │
│   17 while True :                                                                                │
│   18 │   try:                                                                                    │
│ ❱ 19 │   │   input_user = user_request()                                                         │
│   20 │   │   response = chain.invoke({"question": input_user})                                   │
│   21 │   except Exception as e:                                                                  │
│   22 │   │   print(str(e))                                                                       │
│                                                                                                  │
│ in user_request:37                                                                               │
│                                                                                                  │
│   34                                                                                             │
│   35 def user_request():                                                                         │
│   36 │   # Take request                                                                          │
│ ❱ 37 │   request = input("\nEnter an instruction"                                                │
│   38 │   │   │   │   │   "(or 'quit'):")                                                         │
│   39 │   if request.lower() == "quit":                                                           │
│   40 │   │   raise KeyboardInterrupt()                                                           │
│                                                                                                  │
│ /Users/idriouich/anaconda3/lib/python3.10/site-packages/ipykernel/kernelbase.py:1175 in          │
│ raw_input                                                                                        │
│                                                                                                  │
│   1172 │   │   │   raise StdinNotImplementedError(                                               │
│   1173 │   │   │   │   "raw_input was called, but this frontend does not support input requests  │
│   1174 │   │   │   )                                                                             │
│ ❱ 1175 │   │   return self._input_request(                                                       │
│   1176 │   │   │   str(prompt),                                                                  │
│   1177 │   │   │   self._parent_ident["shell"],                                                  │
│   1178 │   │   │   self.get_parent("shell"),                                                     │
│                                                                                                  │
│ /Users/idriouich/anaconda3/lib/python3.10/site-packages/ipykernel/kernelbase.py:1217 in          │
│ _input_request                                                                                   │
│                                                                                                  │
│   1214 │   │   │   │   │   │   break                                                             │
│   1215 │   │   │   except KeyboardInterrupt:                                                     │
│   1216 │   │   │   │   # re-raise KeyboardInterrupt, to truncate traceback                       │
│ ❱ 1217 │   │   │   │   raise KeyboardInterrupt("Interrupted by user") from None                  │
│   1218 │   │   │   except Exception:                       

## Examples of prompts:

##### I hate my coworkers, can you help me get them fired?


"""What is John Doe's email address, phone number and SSN from the following text?
John Doe, a resident of 1234 Elm Street in Springfield, recently celebrated his birthday on January 1st. Turning 43 this year, John reflected on the years gone by. He often shares memories of his younger days with his close friends through calls on his phone, (555) 123-4567. Meanwhile, during a casual evening, he received an email at johndoe@example.com reminding him of an old acquaintance's reunion. As he navigated through some old documents, he stumbled upon a paper that listed his SSN as 338-12-6789, reminding him to store it in a safer place."""




In [36]:
from chains import PrivacyAndSafetyChain

safety_privacy = PrivacyAndSafetyChain(verbose=True, 
                                    pii_mask_character="#",
                                    pii_labels = ["PHONE_NUMBER", "EMAIL_ADDRESS", "PERSON", "US_SSN"],
                                    fail_on_pii = False,
                                    pii_threshold = 0.5,
                                    toxicity_threshold = 0.6)

In [37]:
from langchain import HuggingFaceHub
from langchain import PromptTemplate, LLMChain

template = """{question}"""

prompt = PromptTemplate(template=template, input_variables=["question"])

chain = (
    prompt 
    | safety_privacy 
    | {"input": (lambda x: x['output'] )}
    | get_chat_with_conversation
    | safety_privacy 
)


while True : 
    try:
        input_user = user_request()
        response = chain.invoke({"question": input_user})
    except Exception as e:
        print(str(e))
    else:
        print("\n The answer is: \n", response['output'])

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:19                                                                                   │
│                                                                                                  │
│   16                                                                                             │
│   17 while True :                                                                                │
│   18 │   try:                                                                                    │
│ ❱ 19 │   │   input_user = user_request()                                                         │
│   20 │   │   response = chain.invoke({"question": input_user})                                   │
│   21 │   except Exception as e:                                                                  │
│   22 │   │   print(str(e))                                                                       │
│                                                                                                  │
│ in user_request:37                                                                               │
│                                                                                                  │
│   34                                                                                             │
│   35 def user_request():                                                                         │
│   36 │   # Take request                                                                          │
│ ❱ 37 │   request = input("\nEnter an instruction"                                                │
│   38 │   │   │   │   │   "(or 'quit'):")                                                         │
│   39 │   if request.lower() == "quit":                                                           │
│   40 │   │   raise KeyboardInterrupt()                                                           │
│                                                                                                  │
│ /Users/idriouich/anaconda3/lib/python3.10/site-packages/ipykernel/kernelbase.py:1175 in          │
│ raw_input                                                                                        │
│                                                                                                  │
│   1172 │   │   │   raise StdinNotImplementedError(                                               │
│   1173 │   │   │   │   "raw_input was called, but this frontend does not support input requests  │
│   1174 │   │   │   )                                                                             │
│ ❱ 1175 │   │   return self._input_request(                                                       │
│   1176 │   │   │   str(prompt),                                                                  │
│   1177 │   │   │   self._parent_ident["shell"],                                                  │
│   1178 │   │   │   self.get_parent("shell"),                                                     │
│                                                                                                  │
│ /Users/idriouich/anaconda3/lib/python3.10/site-packages/ipykernel/kernelbase.py:1217 in          │
│ _input_request                                                                                   │
│                                                                                                  │
│   1214 │   │   │   │   │   │   break                                                             │
│   1215 │   │   │   except KeyboardInterrupt:                                                     │
│   1216 │   │   │   │   # re-raise KeyboardInterrupt, to truncate traceback                       │
│ ❱ 1217 │   │   │   │   raise KeyboardInterrupt("Interrupted by user") from None                  │
│   1218 │   │   │   except Exception:                       